### PR TITLE
Multitask Evaluation on Different Magnitude Targets

### PR DESCRIPTION
**Pull request by cjmcgill**
## Description
In multitask models, the metric used for whole model evaluation is an average of the different tasks. This is then used in evaluating the best epoch and for providing the overall test set performance as used in hyperparameter search. The problem is that the metric evaluation is carried out on non-scaled targets, so in regression models the largest magnitude target will dominate for most metrics.

In this PR, I've made the change that when the metric is magnitude-dependent, geometric mean is used instead of arithmetic mean. This bypasses the scale difference by weighing the changes in individual task metrics in a relative sense (10% improvement on task A is worth the same as 10% on task B, no matter the scale).

I've also removed the averaging functions that filter out nan values from the score reporting and the testing functions. If one fold of a model is producing a nan result, it should alert the user to a problem instead of ignoring it.

## Checklist
- [x] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/utils.py [hunk 1]
```diff
@@ -768,3 +770,29 @@ def update_prediction_args(
...
+    :param scores: The scores from different tasks for a single metric.
+    :param metric: The metric used to generate the scores.
+    :axis: The axis along which to take the mean.
+    :return: The combined score across the tasks.
+    """
+    scale_dependent_metrics = ['rmse', 'mae', 'mse', 'bounded_rmse', 'bounded_mae', 'bounded_mse']
+
+    if metric in scale_dependent_metrics:
+        return gmean(scores, axis=axis)
+    else:
```
Comment (jonwzheng) [inline chemprop/utils.py:797]: Might I suggest explicitly including the non-scale dependent metrics here, and raise an error if the metric isn't in either the scale-dependent or non-scale-dependent lists? That way, in a later PR, if someone wants to add a new scale-dependent metric but forgets to update this function, then they get an informative error message.

**Review by jonwzheng — COMMENTED** [review]
Interestingly, I found that 4 years ago the exact opposite change was made to the averaging functions. Meaning all the `np.mean`s were changed to `np.nanmean` in order to "fix bug with multiple folds when some tasks are very inbalanced". Do we have any insight as to why this change was made originally? See https://github.com/chemprop/chemprop/commit/c97773db051594ceb19de9fedc29086ba6947378

File: chemprop/utils.py (hunk 1 above)
Comment (cjmcgill) [reply chemprop/utils.py:797]: This is a good note. I’ll add something like this.

**Comment by cjmcgill** [comment]
@jonwzheng I don't know the specifics, but I can speculate. @yangkevin2 if you're around, getting the actual story would be better than me speculating so please correct me if I miss something.

A lot of the classification datasets are very sparse in positive results. Many of the classification metrics require you to have both positives and negatives to give a meaningful result (and some further require true positives, false positives, true negatives, and false negatives). When doing a many-fold job, some of the splits may randomly have no positives in the test set or validation set, which will lead to weird behavior and return NaN as the metric score for that fold. This problem is even worse for big multitask classification jobs (of which there are many) where the different tasks are sparse in different ways and you have to have them *all* contain positives. When I removed the nanfilters from the integration testing, I had to change the split sizes for some tests from default 80-10-10 to 40-30-30 to avoid this.

Without the nanfilter functions, having a single NaN failure can mess up a whole 10-fold result for just one bad split. So it's reasonable to want to preserve the rest of the models. The problem is, that the nanfilters don't flag the problem. You have a score that evaluates 90% of the models or 90% of the data in a strict crossval, but you think that score represents 100%. Or maybe it's worse, maybe it's only counting 30% of your models because you have a ton of bad splits. So now you have a bundle of models that you want to use to make predictions but you don't know that they are not validated. They could be much worse than you think. So by my thinking, we would be better off going back to not having these nanfilter functions. And users will be responsible for choosing their splits, which they already were they just didn't know the consequences.

This was also a practical matter for this PR because I don't think the scipy geometric mean has a nanfiltered version.

File: chemprop/utils.py (hunk 1 above)
Comment (cjmcgill) [reply chemprop/utils.py]: Okay, added. Let me know what you think.

Comment (jonwzheng) [reply chemprop/utils.py]: Looks good!

**Review by jonwzheng — APPROVED** [review]
The code looks good to me, I will approve this for merging.

I agree that it's undesirable for the code to ignore `nan` s especially if no error is raised. There may be Chemprop users doing this without knowing it (for instance, as you mentioned, you had to modify the splits in the testing module to get a reasonable result), and if the default behavior changes so their approach won't work anymore, this might really confuse them. I think we should also include something to let users know what changed and how to resolve it since the default behavior is changing. Maybe this could be something as simple as an error message if `np.isnan(mean_score)` letting them know that at least one fold contains a null value, probably due to an imbalanced train/test/val split without both positives & negatives [and possibly that the default behavior has changed from the past]. Let me know what you think.

**Comment by cjmcgill** [comment]
Good thinking. I'll add something to that effect.

**Comment by cjmcgill** [comment]
@jonwzheng I added a log message in cross_validate. I didn't make it an error message because there are potentially some places where you would want the output to be nan intentionally, such as when running without a test set. If that log message looks good to you, I will merge.

**Comment by jonwzheng** [comment]
Looks good!
